### PR TITLE
Use Zypak

### PR DESCRIPTION
**Pull request by TheEvilSkeleton**
Closes https://github.com/flathub/com.visualstudio.code/issues/223

### CODE REVIEW

**Comment by TheEvilSkeleton** [comment]
I tried running `flatpak run com.visualstudio.code`, but it doesn't launch and it does not give me an error until I cancel it.

I'm unfamiliar with Python so I don't really know what the issue means here:

```
[📦 com.visualstudio.code ~]$ code 
^CFatal Python error: init_import_size: Failed to import the site module
Python runtime state: initialized
Traceback (most recent call last):
  File "/usr/lib/python3.8/site.py", line 580, in <module>
    main()
  File "/usr/lib/python3.8/site.py", line 573, in main
    execsitecustomize()
  File "/usr/lib/python3.8/site.py", line 512, in execsitecustomize
    import sitecustomize
  File "/usr/lib/python3.8/sitecustomize.py", line 1, in <module>
    from distutils import sysconfig
  File "/usr/lib/python3.8/distutils/sysconfig.py", line 14, in <module>
    import re
  File "/usr/lib/python3.8/re.py", line 127, in <module>
    import functools
  File "/usr/lib/python3.8/functools.py", line 18, in <module>
    from collections import namedtuple
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 779, in exec_module
  File "<frozen importlib._bootstrap_external>", line 911, in get_code
  File "<frozen importlib._bootstrap_external>", line 580, in _compile_bytecode
KeyboardInterrupt
```

**Comment by TheEvilSkeleton** [comment]
I installed `com.visualstudio.code.tool.fish` and `com.visualstudio.code.tool.podman`, and it starting spamming:

```
[📦 com.visualstudio.code ~]$ code 
flatpak-vscode: Adding /app/tools/fish/bin to PATH
flatpak-vscode: Adding /app/tools/podman/bin to PATH
flatpak-vscode: Adding /app/tools/fish/bin to PATH
flatpak-vscode: Adding /app/tools/podman/bin to PATH
flatpak-vscode: Adding /app/tools/fish/bin to PATH
flatpak-vscode: Adding /app/tools/podman/bin to PATH
flatpak-vscode: Adding /app/tools/fish/bin to PATH
flatpak-vscode: Adding /app/tools/podman/bin to PATH
flatpak-vscode: Adding /app/tools/fish/bin to PATH
flatpak-vscode: Adding /app/tools/podman/bin to PATH
flatpak-vscode: Adding /app/tools/fish/bin to PATH
flatpak-vscode: Adding /app/tools/podman/bin to PATH
flatpak-vscode: Adding /app/tools/fish/bin to PATH
flatpak-vscode: Adding /app/tools/podman/bin to PATH
[...]
```

File: code.sh [hunk 1]
```diff
@@ -61,4 +61,4 @@ for i in "${SDK[@]}"; do
 done
 
 exec env PATH="${PATH}:${XDG_DATA_HOME}/node_modules/bin" \
-  /app/extra/vscode/bin/code --extensions-dir=${XDG_DATA_HOME}/vscode/extensions "$@" ${WARNING_FILE}
+  /app/bin/code --extensions-dir=${XDG_DATA_HOME}/vscode/extensions "$@" ${WARNING_FILE}
```
Comment (TheEvilSkeleton) [inline code.sh:64]: Oops, user error :).

**Comment by TheEvilSkeleton** [comment]
What? I'll remove arm64 temporarily.

**Comment by TheEvilSkeleton** [comment]
@gasinvein @barthalion I decided to build locally because flathubbot is acting weird.

I got it working with Zypak:

image

File: code.sh [hunk 2]
```diff
@@ -61,4 +61,4 @@ for i in "${SDK[@]}"; do
 done
 
 exec env PATH="${PATH}:${XDG_DATA_HOME}/node_modules/bin" \
-  /app/extra/vscode/bin/code --extensions-dir=${XDG_DATA_HOME}/vscode/extensions "$@" ${WARNING_FILE}
+  /app/bin/zypak-wrapper.sh /app/extra/code "$@" --extensions-dir=${XDG_DATA_HOME}/vscode/extensions "$@" ${WARNING_FILE}
```
Comment (gasinvein) [inline code.sh:64]: By running the electron binary directly, you completely omit the .js entrypoint responsible for handling cli args.

Comment (TheEvilSkeleton) [reply code.sh:64]: Can you explain what this means?

Anyhow, Zypak is supposed to directly run the binary, and not the wrapper.

Comment (gasinvein) [reply code.sh:67]: > Can you explain what this means?

This means that cli args won't work. Check with `flatpak run com.visualstudio.code --help`. I can't say if anything else will break.

> Zypak is supposed to directly run the binary, and not the wrapper.

Yes, so you'll need to reimplement relevant parts of the upstream wrapper and put in into flatpak's one.

Comment (TheEvilSkeleton) [reply code.sh]: I think I got it working.

Comment (TheEvilSkeleton) [reply code.sh]: Hmm... I thought it worked, but it didn't.

Comment (TheEvilSkeleton) [reply code.sh:67]: @gasinvein do you have an idea how we could reimplement? I can't get it to work.

Comment (gasinvein) [reply code.sh:67]: You've probably forgot `ELECTRON_RUN_AS_NODE=1`.

Comment (gasinvein) [reply code.sh:67]: Also, the really tricky part is passing the cli handler only to the initial process, but not to its reexeced children (zygotes or whatever they're called).
I didn't find any other way to use two wrappers: 1st, "outer", passes the cli handler arg, `--extensions-dir` and `ELECTRON_RUN_AS_NODE`, while the 2nd, "inner", is the normal one that sets up development environment, without passing any additional args; it's the 2nd which is used as `CHROME_WRAPPER`.

I've did mostly the same for Code-OSS, and it mostly works, but still decided to stay with `--no-sandbox` for now. I can push the relevant changes to the git repo if you'd like.

Comment (TheEvilSkeleton) [reply code.sh:67]: > You've probably forgot ELECTRON_RUN_AS_NODE=1.

I forgot to mention that I tried it locally but it didn't work. When I tried without it, I improperly tested it which made me think it worked, but it didn't either, so I just reverted the commit.

> I've did mostly the same for Code-OSS, and it mostly works, but still decided to stay with --no-sandbox for now. I can push the relevant changes to the git repo if you'd like.

What do you mean by relevant changes? Are you talking about my patches?

Comment (gasinvein) [reply code.sh:67]: I mean, I have changes for integrating zypak into code-oss locally, but didn't push them to the flathub git repo.

Comment (TheEvilSkeleton) [reply code.sh:67]: Ah okay. If you can push them in a new branch, that would be nice.

Comment (gasinvein) [reply code.sh:67]: [Here](https://github.com/flathub/com.visualstudio.code-oss/commit/42d4f1baa3d9e17a4cd866556ecb2471b5bdceab) it is.

Comment (TheEvilSkeleton) [reply code.sh:67]: > I've did mostly the same for Code-OSS, and it mostly works, but still decided to stay with --no-sandbox for now. I can push the relevant changes to the git repo if you'd like.

When you say "it mostly works", does it mean that there are some things that do not work or does it mean something else?

Comment (gasinvein) [reply code.sh:67]: From what I noticed, `--wait` option doesn't work. Didn't yet figure out why.

**Pull request closed without merging**